components/service/experiments/src/main/java/mozilla/components/service/experiments/Experiments.kt

```diff
@@ -56,7 +58,8 @@ open class ExperimentsInternalAPI internal constructor() {
      */
     fun initialize(
         applicationContext: Context,
-        configuration: Configuration = Configuration()
+        configuration: Configuration = Configuration(),
+        onExperimentsUpdated: (() -> Unit)? = null
```

Comment:
**Note*: the owners of the experiments library has the final word on this!

While this solves the problem you're seeing (which I guess is the fact that `withExperiment` can be called while the library is not ready, yet!), it also pushes the boilerplate to the user of the API, which now has to deal with the library state (which should be an internal detail).

I think that `withExperiment` should not have a dependency on when `initialize` is called and its dependency should be handled internally, as we do with Glean.

With that said, I think the owner of the library can make the call :)

Reply:
Understood about waiting for approval from the owners of the library.

Experiments in this PR will still work the exact same way it does today, if the callback is not used. The experiment will be received and enrollment determined, only the app won't be able to tell this until the next time it is launched (or if it polls `withExperiment`).  

This adds to the experiments API the ability for an application to receive a notification when the experiments have been updated through the callback, that way it can act on the enrollment without having to wait for launch or polling `withExperiment`.

Reply:
I agree that `withExperiment` should not have an ordering dependency on calls to `initialize`. It would be great if `withExperiment` acted as more of an event listener and less of an if block.

That being said, I think that this change provides a useful improvement to the use cases we are seeing today, and is worth keeping. I think that ultimately we should prefer the improvements to `withExperiment`, or some other reactive experiment checking. Until then, lets take the easy wins that get us unblocked.